Report errors

mstajdohar opened the pull request.

mstajdohar committed:
Save Resolwe errors instead of raising Python exceptions

Check if files exist before saving them.
Add tests for 'save_file', 'save_file_list', 'progress' and 'checkrc'
functions.
Make 'checkrc' and 'progress' function more robust to improper input.

mstajdohar commented:
@dblenkus, I have implemented the [file size task](https://jira.genialis.com/browse/GP-274).

mstajdohar commented:
@tjanez, will you automate the tests on Travis for this repo?

mstajdohar commented:
@dblenkus, do you think file size should be handled by the backend before the Data object is saved to DB?

dblenkus commented:
I prefer handling this in executor. I think that you set here what to save and in executor how to save.

mstajdohar commented:
OK, I will refactor this PR to only include bug fixes and input validation.

mstajdohar commented:
The PR now includes fixes for error reporting. @tjanez, please review.

tjanez commented on resolwe_runtime_utils.py:
```diff
@@ -45,8 +46,18 @@ def save_file(key, file_name, *refs):
     given.
 
     """
+    if not os.path.isfile(file_name):
```
Nice, thanks for implementing file existence checks :+1:

tjanez commented on resolwe_runtime_utils.py:
```diff
@@ -45,8 +46,18 @@ def save_file(key, file_name, *refs):
 
     """
+    if not os.path.isfile(file_name):
+        error('Output \'{}\' missing file: \'{}\''.format(key, file_name))
```
I think this could be worded better. What about `Output \'{}\' has a missing file: \'{}\'`?

mstajdohar replied:
Maybe then `Output \'{}\' is missing a file: \'{}\'` or `Output \'{}\' set to a missing file: \'{}\'` or `File on output \'{}\' does not exist: \'{}\'`?

tjanez replied:
I like the 3rd and the 2nd option. Use whatever you prefer.

tjanez commented on resolwe_runtime_utils.py:
```diff
@@ -45,8 +46,18 @@ def save_file(key, file_name, *refs):
+        missing_refs = [ref for ref in refs if not (os.path.isfile(ref) or os.path.isdir(ref))]
+
+        if len(missing_refs) > 0:
+            error('Output \'{}\' missing references: {}'.format(key, ', '.join(missing_refs)))
```
Similar to above. What about `Output \'{}\' has missing references: \'{}\'`?

tjanez commented on test_resolwe_runtime_utils.py:
```diff
@@ -52,13 +53,17 @@ def test_improper_input(self):
 
 class TestGenSaveFile(ResolweRuntimeUtilsTestCase):
```
If you can, add a test that checks the code when the file or references don't exist.

tjanez commented on test_resolwe_runtime_utils.py:
```diff
@@ -105,11 +110,6 @@ def test_number(self):
     def test_string(self):
         self.assertEqual(progress('0.1'), '{"proc.progress": 0.1}')
 
-    def test_improper_input(self):
```
Please, convert this test to test if we get a platform warning for improper input.

tjanez commented:
@mstajdohar, regarding tests, yes, I intend to test this automatically with Travis.

jkokosar committed:
Add 'save_list' and 'save_file_list' functions and console commands

mstajdohar commented:
@tjanez, I have addressed all comments and merged some features from @jkokosar's PR. Please review ASAP as @jkokosar needs this changes in Docker by the end of this sprint.

tjanez commented on resolwe_runtime_utils.py:
```diff
@@ -45,12 +56,58 @@ def save_file(key, file_name, *refs):
+
+        file_list.append(file_obj)
+
+    return json.dumps({key: file_list})
```
I think it would be better to use `save_list()` here.

tjanez commented:
Please, rebase this on the `master` branch so that Travis CI will run the tests.

Also, Resolwe Runtime Utilities now has [Change Log](https://github.com/genialis/resolwe-runtime-utils/blob/master/docs/CHANGELOG.rst), so add appropriate entries.

mstajdohar commented:
@tjanez, coverage passed (also [coverage of the diff](https://codecov.io/gh/genialis/resolwe-runtime-utils/commit/cffbf210aaccae6a90fb8dcbfeb51846a9df9a2d)), but the integration with codecov broke after some force pushes. Only the Travis test was reported.

mstajdohar commented:
I will try to add another commit (a change log) to see if the codecov integration will reestablish.

mstajdohar committed:
Add 'save_dir' and 'save_dir_list' functions and console commands

Replace 'file_name' with 'file_path' in 'save_file' and 'save_file_list'
functions.

tjanez committed:
Make 'checkrc' and 'progress' functions more robust to improper input

tjanez committed:
Refactor 'save_file_list' function and make it more robust

Add additional tests for 'save_file' and 'save_file_list' functions.

tjanez commented:
@mstajdohar, I've reworked yours and @jkokosar's commits and added two from myself. Please, review it.

mstajdohar commented on test_resolwe_runtime_utils.py:
```diff
@@ -15,9 +15,10 @@
 # pylint: disable=missing-docstring
 import json
 from unittest import TestCase
+from unittest.mock import patch
```
The `unittest` module in Python2 does not include `mock`. Should we support Python2? Which Python do we use in docker?

tjanez replied:
Yes, I know this. I decided to only support Python 3 since it is available in the main "consumer" of Resolwe Runtime Utilities project, the [Resolwe Bioinformatics Docker container](https://github.com/tjanez/docker-bio-linux8-resolwe).

mstajdohar commented:
LGTM

mstajdohar commented:
I have added the `re-save-dir` and `re-save-dir-list` commands that were somehow left out of Tadej's fixes. @tjanez, please review the last commit.

tjanez committed:
Create '_get_json' auxiliary function use it in 'save' and 'save_list'

Thanks to Janez Kokosar for providing the test case for 'save_list'.

tjanez committed:
Refactor 'save_dir_list' function

Add additional tests for 'save_dir' and 'save_dir_list' functions.
